File: tests/common/positive/systems/multiple_handlers.rb

class SFrameEvent
    attr_accessor :_message
    attr_accessor :_parameters

    def initialize(message, parameters = {})
        @_message = message
        @_parameters = parameters
    end
end


class SFrameContext
    attr_accessor :_event
    attr_accessor :_return
    attr_accessor :_data

    def initialize(event, default_return = nil)
        @_event = event
        @_return = default_return
        @_data = {}
    end
end


class SCompartment
    attr_accessor :state
    attr_accessor :state_args
    attr_accessor :state_vars
    attr_accessor :enter_args
    attr_accessor :exit_args
    attr_accessor :forward_event
    attr_accessor :parent_compartment

    def initialize(state, parent_compartment = nil)
        @state = state
        @state_args = {}
        @state_vars = {}
        @enter_args = {}
        @exit_args = {}
        @forward_event = nil
        @parent_compartment = parent_compartment
    end

    def copy
        c = SCompartment.new(@state, @parent_compartment)
        c.state_args = @state_args.dup
        c.state_vars = @state_vars.dup
        c.enter_args = @enter_args.dup
        c.exit_args = @exit_args.dup
        c.forward_event = @forward_event
        c
    end
end


class S
    attr_accessor :_state_stack
    attr_accessor :__compartment
    attr_accessor :__next_compartment
    attr_accessor :_context_stack

    def initialize
        @_state_stack = []
        @_context_stack = []
        # HSM: Create parent compartment chain
        __parent_comp_0 = SCompartment.new("P", nil)
        @__compartment = SCompartment.new("A", __parent_comp_0)
        @__next_compartment = nil
        __frame_event = SFrameEvent.new("$>")
        __ctx = SFrameContext.new(__frame_event, nil)
        @_context_stack.push(__ctx)
        __kernel(@_context_stack[@_context_stack.length - 1]._event)
        @_context_stack.pop
    end

    def __kernel(__e)
        # Route event to current state
        __router(__e)
        while @__next_compartment != nil
            next_compartment = @__next_compartment
            @__next_compartment = nil
            exit_event = SFrameEvent.new("<$", @__compartment.exit_args)
            __router(exit_event)
            @__compartment = next_compartment
            if next_compartment.forward_event == nil
                enter_event = SFrameEvent.new("$>", @__compartment.enter_args)
                __router(enter_event)
            else
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = nil
                if forward_event._message == "$>"
                    __router(forward_event)
                else
                    enter_event = SFrameEvent.new("$>", @__compartment.enter_args)
                    __router(enter_event)
                    __router(forward_event)
                end
            end
        end
    end

    def __router(__e)
        state_name = @__compartment.state
        handler_name = "_state_#{state_name}"
        if respond_to?(handler_name, true)
            send(handler_name, __e)
        end
    end

    def __transition(next_compartment)
        @__next_compartment = next_compartment
    end

    def e1
        __e = SFrameEvent.new("e1", {})
        __ctx = SFrameContext.new(__e, nil)
        @_context_stack.push(__ctx)
        __kernel(__e)
        @_context_stack.pop
    end

    def e2
        __e = SFrameEvent.new("e2", {})
        __ctx = SFrameContext.new(__e, nil)
        @_context_stack.push(__ctx)
        __kernel(__e)
        @_context_stack.pop
    end

    def _state_P(__e)
    end

    def _state_A(__e)
        if __e._message == "e1"
            _state_P(__e)
        elsif __e._message == "e2"
            _state_P(__e)
        end
    end
end

puts("TAP version 14")
puts("1..1")
begin
    s = S.new
    if s.respond_to?(:e)
        s.e()
    end
    puts("ok 1 - multiple_handlers")
rescue => ex
    puts("not ok 1 - multiple_handlers # #{ex}")
end

File: tests/common/positive/systems/multiple_handlers.php
from typing import Any, Optional, List, Dict, Callable

<?php

class SFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class SFrameContext:
    def __init__(self, event: SFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class SCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'SCompartment':
        c = SCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class S:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        # HSM: Create parent compartment chain
        __parent_comp_0 = SCompartment("P", parent_compartment=None)
        self.__compartment = SCompartment("A", parent_compartment=__parent_comp_0)
        self.__next_compartment = None
        __frame_event = SFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = SFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = SFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = SFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def e1(self):
        __e = SFrameEvent("e1", None)
        __ctx = SFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def e2(self):
        __e = SFrameEvent("e2", None)
        __ctx = SFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def _state_P(self, __e):
        pass

    def _state_A(self, __e):
        if __e._message == "e1":
            self._state_P(__e)
        elif __e._message == "e2":
            self._state_P(__e)

function native() {}
function x() {}

echo "TAP version 14" . "\n";
echo "1..1" . "\n";
try {
    $s = new S();
    if (method_exists($s, 'e')) { $s->e(); }
    echo "ok 1 - multiple_handlers" . "\n";
} catch (\Exception $ex) {
    echo "not ok 1 - multiple_handlers # " . $ex->getMessage() . "\n";
}

